### PR TITLE
[browser][MT] fix stale memory on suspended thread

### PR DESCRIPTION
**Pull request by pavelsavara**
Alternative to https://github.com/dotnet/runtime/pull/90771

Problem: when WebWorker is suspended in the browser, the other running threads could `grow` the linear memory in the meantime. After the thread is un-suspended it may to try to de-reference pointer which is beyond it's known view. This is likely V8 bug.

We believe that described problem is cause of MT problems with debugger tests, example [log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-93920-merge-f5727b81f7c54bb190/chrome-DebuggerTests.PointerTests/1/console.06520613.log?helixlogtype=result)

### CODE REVIEW

**Comment by pavelsavara** [comment]
This helps
image

File: src/mono/wasm/runtime/memory.ts
Comment (ilonatommy) [inline src/mono/wasm/runtime/memory.ts:414]: Why false when we never check the returned value?
```suggestion
    if (!MonoWasmThreads) return;
```

File: src/mono/wasm/runtime/memory.ts
Comment (ilonatommy) [inline src/mono/wasm/runtime/memory.ts:417]: Without updating the effect does not get propagated? What is the role of update exactly?

**Review by ilonatommy — APPROVED** [review]

File: src/mono/wasm/runtime/memory.ts
Comment (ilonatommy) [reply src/mono/wasm/runtime/memory.ts:417]: Ok, I see, it's how they do it, the re-allocate the heaps: https://github.com/emscripten-core/emscripten/blob/a2f3b925c967b9a8091b5b52fa8cbb649cfe248e/src/preamble.js#L146

Comment (pavelsavara) [reply src/mono/wasm/runtime/memory.ts:417]: `wasmMemory.grow` always invalidates existing views

File: src/mono/wasm/runtime/memory.ts
Comment (kg) [inline src/mono/wasm/runtime/memory.ts:422]: Please add a comment explaining the significance of grow(0) and why we're doing it, i.e.

"Normally when wasm memory grows in v8, this size change is broadcast to other web workers via an 'interrupt', which works by setting a thread-local flag that needs to be checked. It's possible that at this point in execution the flag has not been checked yet (because this worker was suspended by the debugger in an unknown location), which means the size change has not taken effect in this worker. wasmMemory.grow's implementation in v8 checks to see whether other workers have already grown the buffer, and will update the current worker's knowledge of the buffer's size. After that we should be able to safely updateMemoryViews and get a correctly sized view. This only works because their implementation does not skip doing work even when you ask to grow by 0 pages."

**Comment by pavelsavara** [comment]
@thaystg I will wait for your feedback/approval

**Comment by pavelsavara** [comment]
Hmm, so now all errors are `"TypeError: Failed to fetch\n    at Object.L [as fetch_like] (http://localhost:9400/_framework/dotnet.js:3:4721)\n    at http://localhost:9400/_framework/dotnet.js:3:12002\n    at http://localhost:9400/_framework/dotnet.js:3:12018"` [Log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-94299-merge-c1b09c0202b6489e91/chrome-DebuggerTests.AsyncTests/1/console.0ca4bc40.log?helixlogtype=result)


@thaystg does that mean my fix didn't help ? Or it means there is another issue ?

Is debugger making the app to download something ?

Does suspending thread while fetching break the fetch ?

**Comment by thaystg** [comment]
There are a lot of core dumps generated: chrome is crashing, not sure if it's related to your fix.

image

**Comment by thaystg** [comment]
I also remember to have already tried it and it didn't help. But for me it's okay to try again. :)

**Comment by pavelsavara** [comment]
> I think there are a lot of core dumps generated: chrome is crashing, not sure if it's related to your fix.

Where is "there" ?

More details would help.

> I also remember to have already tried it and it didn't help. But for me it's okay to try again. :)

Could you please elaborate on what you tried ?

**Comment by thaystg** [comment]
> > I think there are a lot of core dumps generated: chrome is crashing, not sure if it's related to your fix.
> 
> Where is "there" ?
> 
> More details would help.
> 
> > I also remember to have already tried it and it didn't help. But for me it's okay to try again. :)
> 
> Could you please elaborate on what you tried ?

Look at artifacts tab here: https://dev.azure.com/dnceng-public/public/_build/results?buildId=459235&view=ms.vss-test-web.build-test-results-tab&runId=10361536&resultId=100252&paneView=dotnet-dnceng.dnceng-build-release-tasks.helix-test-information-tab

image

**Comment by thaystg** [comment]
> Could you please elaborate on what you tried ?

Unfortunately I don't remember details of what I tested, I tried a lot of things before the workaround in that PR. 
If I can help you in anyway by testing your solution or anything else, please let me know.

**Comment by pavelsavara** [comment]
image

**Comment by pavelsavara** [comment]
I think that the fix for memory views is good. Solid 10/10 
image


There are problems with https://github.com/dotnet/runtime/issues/89409 but that's unrelated issue
```
The active test run was aborted. Reason: Test host process crashed
Results File: /root/helix/work/workitem/uploads/xharness-output/logs/testResults.trx

Passed!  - Failed:     0, Passed:    80, Skipped:     0, Total:    80, Duration: 9 m - DebuggerTestSuite.dll (net8.0)
Test Run Aborted.
```

**Comment by thaystg** [comment]
@pavelsavara thanks a lot and congrats for this fix!!!

**Pull request merged**
